rfc-120-csv-preview.md

```diff
@@ -0,0 +1,134 @@
+
+   - A 'header' section with details of the parent document (and navigation links)
+   - A 'body' section with a preview of the content of the CSV attachment
+
+This CSV Preview feature only works for CSV file attachments uploaded through Whitehall. We need it to work for CSV attachments published through Content Publisher, which is superseding Whitehall.
+
+## Proposal
+
+### Step 1
+
+Create a new `attachment_preview` format in [govuk-content-schemas][] to represent the preview page for an attachment. A publishing app will publish content items in this format for each CSV attachment to preview.
+
+```
+# formats/attachment_preview.jsonnet
+(import "shared/default_format.jsonnet") + {
+  document_type: "attachment_preview",
+  attachment: {
+    "$ref": "#/definitions/file_attachment_asset"
+  },
+  edition_links: {
+    parent: {
+      description: "The parent content item.",
+      maxItems: 1,
+    },
+  }
+}
+```
+
+- Using a content item is consistent with the one-to-one mapping of pages and content on GOV.UK. This should make it easier to debug issues with the feature, since it follows an existing pattern.
+
+- The schema includes a link to the parent document. Using an expanded link means we don't need to make an extra API call to the Content Store in order to show information about the parent document.
+
+- Using the publishing system means we can follow an existing approach for things like automatic redirects (in case the base path changes) and anonymous preview ('fact check'). However, this entails an overhead of extra API calls to manage the separate content items, with the risk they get out-of-sync.
+
+- Using a separate content item means we're not coupled to a particular rendering app. This provides a foundation for extending the preview feature in the future.
+
+### Step 2
+
+Re-implement the CSV Preview feature in Government Frontend.
+
+- Government Frontend currently renders all documents published by Whitehall, that have CSV file attachments. This makes it a relevant codebase to implement this feature in.
+
+- We expect the implementation to be small, isolated and not to grow/change much. Adding a new responsibility to Government Frontend seems like a reasonable compromise when compared to the [rejected proposals](#rejected-proposals).
+
+
+### Step 3
+
+Change Whitehall to use the new CSV Preview feture for newly published documents. We will need to use a different URL structure, since the current one is coupled to Whitehall.
+
+```
+# current URL example
+https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/856932/Management_information_-_further_education_and_skills_-_as_at_31_December_2019.csv/preview
+
+# proposed URL example
+https://www.gov.uk/government/remaining-parent-document-base-path/preview/asset-id
```

Comment:
Could you explain here why the `remaining-parent-document-base-path` portion of the URL is necessary? I didn't get my answer until the end of the document:

> - It would be hard to reproduce the 'header' section of the preview page, since Asset Manager does not store data about the parent document

Have we run this proposal past publishers? I think we may have made an assumption that using the parent document slug and asset ID is acceptable in lieu of a filename, but publishers may not be happy that the filename is no longer reflected in the URL. I don't see this explained/justified in the RFC at the moment.

I can also imagine cases where the publisher might not want the asset to be coupled to a particular document. We may have two different pages referring to the same asset - under the current proposal, one of their slugs would be in the asset preview URL, and one would not.

The 'header' section certainly complicates things, but we can deduce what should go in there by looking at the parent document content item - I don't think the document necessarily needs to be part of the asset preview URL.

For feature parity with the current implementation, I think I would opt for a structure like this: `https://www.gov.uk/government/uploads/preview/asset-id/asset-filename.csv`. If the publisher renames the file (either explicitly or indirectly through a file replacement), this should delegate to the standard redirect patterns we have in publishing API / content store.

Keeping it under `/government/uploads/preview` would involve [adding another regex entry to govuk-puppet](https://github.com/alphagov/govuk-puppet/blob/137e774faa1eb3ff1122c6e71645bcfcfef7815f/hieradata/common.yaml#L372-L377), but I don't see a particular disadvantage to that.

Reply:
The nice thing about this proposal is that it decouples the URL from the implementation; we can indeed use the expanded `parent` link to help us render the page header.

I think the general choice of URL is subjective, and not something we need agreement from publishers on. Publishers aren't able to modify the URLs currently, and I'm not aware of any scenario where this has been a real issue.

If we were starting from scratch today, I don't think we would have come up with `/government/uploads/preview/asset-id/asset-filename.csv`. There are a few objective reasons for not using this particular structure:

- It's semantically coupled to Whitehall: `/government/uploads` is a Whitehall thing, which we're trying to move away from, with the `assets.` domain.

- It's less stable: the URL contains two IDs, which could both change. Only one ID is required, of which the `asset-id` is more stable.

- It's more surprising: [these kinds of ad-hoc regex overrides](https://github.com/alphagov/govuk-puppet/blob/137e774faa1eb3ff1122c6e71645bcfcfef7815f/hieradata/common.yaml#L372-L377) are part of what makes GOV.UK routing (even) hard(er) to understand.

- It's inconsistent: the proposal we have here is to treat CSV Preview like any other GOV.UK content, so it should have a normal-looking URL structure for content.

The proposed URL structure was self-explanatory to me, with the exception of the choice of `/preview/`. We do have a precedent for it: both HTML attachments and manual sections have are 'child' documents that (redundantly) have the parent in the base path.

Does that help? I can add some or all of this into the RFC if need be.

Reply:
> Publishers aren't able to modify the URLs currently, and I'm not aware of any scenario where this has been a real issue.

We've had tickets on 2nd line asking to rename assets, e.g. `court-hearing-22-april` => `court-hearing-23-april`, which seems a reasonable request if they originally made a mistake when naming the asset. I like that under your proposal this will never be an issue, as the filename is never exposed.

I think it would be worth adding an explanation for the proposed URL structure, so that we have something we can point to in the future, and so that it shows you've thought about it. This should include an explanation for the `remaining-parent-document-base-path` portion of the URL - as we've not coupled files to their parent document in the past - and explain why we're not just using the `parent` link in the content ID to get that information for the header.

Reply:
I'm happy to write something. Can you help me understand what the explanation should look like? The discussion we're having here is really about why the original URL structure wouldn't be appropriate, but doesn't preclude other structures. Would be it be sufficient to document why we're not using the original URL structure?